php/tests/php_implementation_test.php

```diff
@@ -513,4 +513,128 @@ public function testPackedByteSize()
+            'map_sint64_sint64' => [-65 => -65],
+            'map_fixed32_fixed32' => [66 => 66],
+            'map_fixed64_fixed64' => [67 => 67],
+            'map_sfixed32_sfixed32' => [-68 => -68],
+            'map_sfixed64_sfixed64' => [-69 => -69],
+            'map_int32_float' => [1 => 3.5],
+            'map_int32_double' => [1 => 3.6],
+            'map_bool_bool' => [true => true],
+            'map_string_string' => ['e' => 'e'],
+            'map_int32_bytes' => [1 => 'f'],
+            'map_int32_enum' => [1 => TestEnum::ONE],
+            'map_int32_message' => [1 => ['a' => 36]],
+        ]);
+
+        TestUtil::assertTestMessage($m);
+
+        // Using message objects
+        $m = new TestMessage([
+            'optional_message' => new TestMessage_Sub(['a' => 33]),
+            'repeated_message' => [
+                new TestMessage_Sub(['a' => 34]),
+                new TestMessage_Sub(['a' => 35]),
+            ],
+            'map_int32_message' => [
+                1 => new TestMessage_Sub(['a' => 36])
+            ],
+        ]);
+
+        $this->assertEquals(33, $m->getOptionalMessage()->getA());
+        $this->assertEquals(34, $m->getRepeatedMessage()[0]->getA());
+        $this->assertEquals(35, $m->getRepeatedMessage()[1]->getA());
+        $this->assertEquals(36, $m->getMapInt32Message()[1]->getA());
+
+        // Using both arrays and objects
+        $m = new TestMessage([
+            'repeated_message' => [
+                new TestMessage_Sub(['a' => 34]),
+                ['a' => 35],
+            ],
+        ]);
+
+        $this->assertEquals(34, $m->getRepeatedMessage()[0]->getA());
+        $this->assertEquals(35, $m->getRepeatedMessage()[1]->getA());
+    }
+
+    public function testMergeFromArrayWithNullValuesDoesNotThrowException()
+    {
+        $requestData = [
+            'optional_int32' => null,
+            'optional_int64' => null,
+            'optional_uint32' => null,
+            'optional_uint64' => null,
+            'optional_sint32' => null,
+            'optional_sint64' => null,
+            'optional_fixed32' => null,
+            'optional_fixed64' => null,
+            'optional_sfixed32' => null,
+            'optional_sfixed64' => null,
+            'optional_float' => null,
+            'optional_double' => null,
+            'optional_bool' => null,
+            'optional_string' => null,
+            'optional_bytes' => null,
+            'optional_enum' => null,
+            'optional_message' => null,
+        ];
+
+        $messageFromArray = new TestMessage($requestData);
+        $messageFromJson = new TestMessage();
+        $messageFromJson->mergeFromJsonString(json_encode($requestData));
+
+        foreach ($requestData as $key => $value) {
+            $getter = 'get' . str_replace(' ', '', ucwords(
+                str_replace('_', ' ', $key)
+            ));
+            $this->assertTrue(empty($messageFromArray->$getter()));
```

Comment:
Primitive field default is not null. e.g., 0 for int32

Reply:
The function `empty` evaluates to true for zero also:
```php
php > var_dump(empty(0));
bool(true)
```

Reply:
Setting fields with null is undefined behavior. Don't test it.

Reply:
Do you want me to remove this test entirely then?

Reply:
I think you only need to test optional message